### DISCUSSION TITLE
Distributed transactions support

### DISCUSSION THREAD

**Discussion by dumblob** [ravendb/ravendb#13216] [Q&A]
I'm a bit puzzled about the degree to which ravendb supports distributed transactions and what are the subtle properties and (dis)advantages. Could you elaborate on that a bit?

**Comment by ayende** [comment]
Distributed transactions in RavenDB are called Cluster Wide transactions, see the discussion on the current behavior here:
https://ravendb.net/articles/ravendb-5-2-simplifying-atomic-cluster-wide-transactions

They are run through the Raft cluster and ensured consistency of operations

**Comment by dumblob** [comment]
Thanks! Looks promising.

What's the performance of these cluster-wide atomic transactions? Any benchmark results to look at?

**Reply by ayende** [reply]
Depending on the actual size of the transaction, typically, about 3 times the ping time in the network.

**Reply by dumblob** [reply]
For each distributed transaction? Is there a way to mitigate this rather cost (e.g. by up-front compilation of a transaction and later just calling the compiled & in-runtime-maintained-based-on-cluster-dynamics plan)?

**Reply by ayende** [reply]
For each distribution transaction, yes. But note that concurrent distributed transactions will be batches and the cost isn't linear. 

The *reason* for this cost is that a distributed transaction requires a consensus. Basically, you have to have a majority of the nodes agreeing on the state of the transaction, that means that we need to typically contact at least another node, then apply the transaction itself.

Multiple concurrent transactions can be pushed to other nodes for confirmations, but the cost above is about sequential transactions, where you wait for each transaction to commit before hand.
Note that distributed transactions also have about 3 times the cost in the number of disk writes. (Once to actually commit the transaction, once to log the transaction in the source server and once to log it in another, assuming 3 node cluster).

This has nothing to do with the actual cost of executing the transaction, given network & I/O, I'm not even counting that cost.

**Reply by dumblob** [reply]
>but the cost above is about sequential transactions, where you wait for each transaction to commit before hand.

This is also what I meant, so we're on the same boat :wink:.

>Note that distributed transactions also have about 3 times the cost in the number of disk writes. (Once to actually commit the transaction, once to log the transaction in the source server and once to log it in another, assuming 3 node cluster).
>
>This has nothing to do with the actual cost of executing the transaction, given network & I/O, I'm not even counting that cost.

Yep, that's what I anticipated when I saw you're using raft.

>For each distribution transaction, yes. But note that concurrent distributed transactions will be batches and the cost isn't linear.

Now this gets interesting. Could you point me to a description how the batching works, what are the heuristics, prioritization, etc.?

**Reply by ayende** [reply]
This is a natural consequence of a how Raft works.
Whenever we send `AppendEntries`, we send multiple entries, so we can process multiple transactions at the same time. 

When we commit, we are also working in batches, so we optimize both I/O & Network in this regard.

I talk about this here in more details, see:
https://ayende.com/blog/174945/the-guts-n-glory-of-database-internals-merging-transactions

**Reply by dumblob** [reply]
> When we commit, we are also working in batches, so we optimize both I/O & Network in this regard.
> 
> I talk about this here in more details, see: https://ayende.com/blog/174945/the-guts-n-glory-of-database-internals-merging-transactions

Yep, this "trick" is well-known (at least to me). But except for optimizing for throughput under high load, do you also optimize for latency in case of low load?

**Reply by ayende** [reply]
Not sure that I'm following you here. We aren't using Nagle or similar, if that is the intent.
As soon as we have *any* work, we'll execute it, and any work that happens during that processing will happen in the next batch

**Reply by dumblob** [reply]
>As soon as we have any work, we'll execute it, and any work that happens during that processing will happen in the next batch

I mean solely the "batching" itself (nothing with network). In other words to make batching work you have to determine what is a batch. Assume there is currently no batch in working (the system is fully idle), then usually batching happens by queing the first incoming transaction and waiting for a certain time for other incoming transactions up to some max `M`. This "waiting" is what I meant because under 100% load there is no waiting, it's just "taking immediately as many transactions up to `M`" (max throughput) but under 1% load there is this waiting which then determines the minimum latency. It's silly to make this waiting to be fixed (e.g. a fixed amount of time or fixed number of transactions etc.).

So my question is how do you modulate this waiting to avoid the fixed nature? Do you try to estimate the current load and the distribution of incoming transactions in time to dynamically determine how much waiting is necessary?

**Reply by karmeli87** [reply]
Hi, 
What you have described is more or less nagle's algorithm and we aren't using it.

**Reply by dumblob** [reply]
Nagle's algorithm is about batching when sending out transactions/packets (i.e. client-side). What I described is about batching when receiving transactions/packets (i.e. server-side).

But if you're not using even this, then how do you determine when to stop buffering into the current batch at the server side and begin with actual processing the batched transactions?

With this misunderstanding I wonder whether you even batch at the server side?

**Reply by karmeli87** [reply]
We send JSON over TCP, once our parser read successfully a JSON entity, we start processing it. In this context a single JSON is a batch that may contain several raft commands.

The batching on the receiving side is dominated by both disk I/O & network. The leader send us commands that we should append/commit and we start processing them, in the meanwhile the leader will accumulate more commands to send us once we are done.

**Reply by dumblob** [reply]
>The batching on the receiving side is dominated by both disk I/O & network. The leader send us commands that we should append/commit and we start processing them, in the meanwhile the leader will accumulate more commands to send us once we are done.

So I (one or very few rogue or badly written clients) can quite easily achieve about 20x less performance of your cluster by carefully timing individual JSON requests (transactions) to make the leader dispatch them individually (i.e. without accumulating) to nodes and thus kind of force the cluster process them one by one by working around all potential batching.

Or is there any auto-tuning of batching going on?

**Reply by ayende** [reply]
I wrote about this in more detail here:

https://ayende.com/blog/195778-C/optimizing-local-and-distributed-transactions-with-batching?key=d9cc771eef7c48af9d97d449deff0097

**Reply by ayende** [reply]
To answer your question:

Note that we don't consider byzantine mode for our system, you are assume to be trusted (once you have authenticated).

That said, I'm not sure that I'm following you. A client that sends requests one at a time would have them processed one at a time, sure, but that won't impact _other_ clients.
We'll merge operations from multiple clients, and a single client sending a request on its own time won't impact other clients.

**Reply by karmeli87** [reply]
Sure you can misuse the system, but why would you want to do that :-)

You can't control the batching. If you really like you can maybe queue those commands on your side and send them at once to the leader.
Can I ask if there is something specific that you are trying to achieve with raven? or simply to better understand how cluster-wide transaction in raven work?

**Reply by dumblob** [reply]
@ayende I really appreciate the interest and speed you're responding with. I'm truly amazed. I feel RavenDB ecosystem is very healthy which is a very good sign. Thank you! Now let's get back to the technical part :wink:.

The post https://ayende.com/blog/195778-C/optimizing-local-and-distributed-transactions-with-batching?key=d9cc771eef7c48af9d97d449deff0097 explains most of the things I was looking for. I'll ask about some more clarifications.

a) Essentially you have two queue instances - one being filled with incoming transactions and the other being read & merged by the parallel thread (or alternatively you use some form of a ring buffer which is the closest-to-optimal solution I know of). Correct? If not, could you clarify how do you achieve the incoming transactions not being rejected at the moment when the whole queue instance is being locked by the read & merge thread?

b) What are the limits on a batch the merge thread can emit under the highest possible load? Is it some cumulative size together with some timeout (otherwise affected clients might starve)? Or is it more complicated (maybe some prioritization, shuffling, ...)?

Let me apologize for the word "rogue". By that I simply meant a properly authenticated client which will have a "badly written" use case and will saturate the cluster on its own. But that's not a big deal as based on my revised understanding it doesn't affect other clients by any means and we're not optimizing for power consumption this extremely :wink:. So no byzantine mode expectations/needs from my side.

>Can I ask if there is something specific that you are trying to achieve with raven? or simply to better understand how cluster-wide transaction in raven work?

Both :wink:. I'd like to achieve minimum possible latency (allowed by the given HW and network) for each client individually in all possible scenarios while avoiding starvation (i.e. guarantee fixed maximum latency). And of course, I'd like to understand how cluster-wide transactions are implemented before diving into the sources.

Actually for a few years I have one crazy idea in my head about cluster- and actually earth- wide transaction processing which I couldn't find implemented anywhere and which could (maybe, who knows :wink:) lower the latencies imposed by consensus-based transactions without sacrificing any of the guarantees (serializability or even strict serializability). I'm pretty sure I won't have the time to implement it myself but I'm open to passing the idea to someone interested and working together on making it happen.

**Reply by ayende** [reply]
a) There is just a single queue, not two of them. The queue is a parallel queue, which allows concurrent enqueue & dequeue. 
I don't think that I follow what you mean there? 

b) The transaction merger will merge transactions until any of the following conditions happen (Controlled by the `TransactionMerger` config):

* Transaction size (defaults to min of 10% of memory or 512MB) 
* There are no more transactions to merge

We'll start rejecting requests from the transaction queue if we have > 512 MB of modified memory and the previous transaction is still doing async commit and didn't complete (to reduce memory load).

The size is the cumulative size of the modified data.
